### PR TITLE
TRM labs integration

### PR DESCRIPTION
**Pull request by piekczyk**
# [TRM labs integration](https://app.shortcut.com/oazo-apps/story/3817/integrate-trm-labs-api)

<please insert a clubhouse link above>
  
## Changes 👷‍♀️
  <Please add short list of changes>

- added risk verification after ToS flow
  
## How to test 🧪
  <Please explain how to test your changes>

- use different kind of wallets and addresses and check whether risk is verified within network tab
- check in db whether additional information are stored in tos_approval (chain_id, signature, message)
- on goerli wallet_risk table in db shouldn't be updated as for test network we always return `isRisky: false`
- on mainnet wallet_risk table should be updated
    
## Definition of done ✔️

- [ ] Acceptance criteria for each issue met
- [ ] Unit tests written where needed and passing
- [ ] End-to-end tests for happy path
- [ ] Documentation updated <When applicable>
- [ ] Project builds without errors
- [ ] Non-functional requirements met
- [ ] Code reviewed and functionality tested by the reviewer (locally, Heroku, etc.)
- [ ] Feature verified and accepted by Product Owner
- [ ] Project deployed to production environment

### CODE REVIEW

**Review by johnnieskywalker — COMMENTED** [review]
Shouldn't this feature be controlled by flag in features ?

**Comment by johnnieskywalker** [comment]
> Shouldn't this feature be controlled by flag in features ?

Changed my mind, this would be serious security vulnerability

**Comment by johnnieskywalker** [comment]
It worked hh addresses are unallowed on mainnet 
image

**Comment by johnnieskywalker** [comment]
Gnosis 
image

testing with goerli url 

https://gnosis-safe.io/app/gor:0x9Cb65f545b0FB0bFEa3254bD08989Fd5Fc6Cdf42/apps?appUrl=https://localhost:3443/?network=goerli

image

image


image


image

image

image

Everything tested and works

**Review by johnnieskywalker — DISMISSED** [review]
Both features work, code is properly organized, approval granted

**Review by johnnieskywalker — DISMISSED** [review]
Good idea

File: features/termsOfService/jwt.ts [hunk 1]
```diff
@@ -12,24 +13,40 @@ export type JWToken = string
 
 export function jwtAuthGetToken(address: string): JWToken | undefined {
   const token = localStorage.getItem(`token-b/${address}`)
+  if (token && token !== 'xxx') {
```
Comment (paszkowskiDamian) [inline features/termsOfService/jwt.ts:16]: `xxx` is it some leftover after tests?

File: features/termsOfService/jwt.ts [hunk 2]
```diff
@@ -12,24 +13,40 @@ export type JWToken = string
...
+async function requestJWT(web3: Web3, account: string, isGnosisSafe: boolean): Promise<string> {
+  const web3Instance = isGnosisSafe ? new Web3(web3.givenProvider) : web3
+  const addressForSignature = isGnosisSafe ? web3Instance.givenProvider.selectedAddress : account
   const challenge = await requestChallenge(account).toPromise()
 
   console.log('Signing challenge: ', challenge)
 
-  const signature = await signTypedPayload(challenge, web3, account)
-  const jwt = await requestSignin({ challenge, signature }).toPromise()
+  const chainId = getNetworkId()
```
Comment (paszkowskiDamian) [inline features/termsOfService/jwt.ts:46]: can you get networkId from `web3Instance`? The function will be more pure then

File: handlers/risk/get.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    riskScoreLevel: number
+    riskScoreLevelLabel: string
+    trmAppUrl: string
+    trmUrn: string
+  }[]
+  trmAppUrl: string
+}
+
+async function getTrmRisk(account: string): Promise<RiskDataResponse> {
+  return fetch('https://api.trmlabs.com/public/v2/screening/addresses', {
```
Comment (paszkowskiDamian) [inline handlers/risk/get.ts:43]: I think `fetch` is not available on server. Did you test this endpoint?

File: handlers/risk/get.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  return !!trmData.addressRiskIndicators.length
+}
+
+const inputSchema = z.object({
+  chainId: z.number(),
+})
+
+export async function getRisk(req: NextApiRequest, res: NextApiResponse) {
+  // TODO provide correct typing after Damians changes
```
Comment (paszkowskiDamian) [inline handlers/risk/get.ts:81]: ❤️

File: handlers/risk/get.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  const user = getUserFromRequest(req as any)
+  const body = inputSchema.parse(req.body)
+
+  if (body.chainId !== 1) {
+    return res.status(200).json({ isRisky: false })
+  }
```
Comment (paszkowskiDamian) [inline handlers/risk/get.ts:87]: I think you can disregard chainId, I'm afraid that it is possible to stub body parameters. 

If the connected address is risky (on mainnet) just don't allow them to use app at all 🤷‍♂️

File: server/app.ts [hunk 6]
```diff
@@ -22,28 +20,9 @@ export function getApp(config: Config, { nextHandler }: Dependencies): express.A
...
-      next()
-    })
-  }
-
   app.use(bodyParser.json())
   app.use(cookieParser())
 
-  if (!config.disableRequestLogging) {
-    app.use('/api', morgan('tiny'))
-  }
```
Comment (paszkowskiDamian) [inline server/app.ts:45]: let's check if we using this logs ;) I know I've removed it in my PR but I'm having second thoughts.

**Review by adamskrodzki — APPROVED** [review]
tested and works

File: features/walletAssociatedRisk/walletRiskApi.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import { catchError, map } from 'rxjs/operators'
+
+const basePath = getConfig()?.publicRuntimeConfig?.basePath || ''
+
+export interface WalletRiskResponse {
+  isRisky?: boolean
+  error?: string
+}
+
+export function getWalletRisk$(token: string, chainId: number): Observable<WalletRiskResponse> {
```
[the remark this answers is not in the document]
Comment (piekczyk) [reply features/walletAssociatedRisk/walletRiskApi.ts:14]: address is decoded from token

File: features/walletAssociatedRisk/WalletAssociatedRisk.tsx [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+import { useAppContext } from '../../components/AppContextProvider'
+import { disconnect } from '../../components/connectWallet/ConnectWallet'
+import { useObservable } from '../../helpers/observableHook'
+
+interface WithWalletAssociatedRiskProps {
+  children: ReactNode
+}
+
+export function WithWalletAssociatedRisk({ children }: WithWalletAssociatedRiskProps) {
```
[the remark this answers is not in the document]
Comment (piekczyk) [reply features/walletAssociatedRisk/WalletAssociatedRisk.tsx:12]: due to dependency on ToS flow

**Review by adamskrodzki — APPROVED** [review]

**Comment by paulmillr** [comment]
@adamskrodzki @paszkowskiDamian @piekczyk why are you doing this? Do you live in the U.S.?

Bootlicking did not help Tornado. They had compliance section and still got banned. So what's the purpose?

**Pull request merged**